src/BlockTypes/ProductStockIndicator.php

```diff
@@ -48,7 +50,88 @@ protected function get_block_type_supports() {
+	 */
+	protected function get_block_type_uses_context() {
+		return [ 'query', 'queryId', 'postId' ];
+	}
+
+	/**
+	 * Get stock text based on stock. For example:
+	 * - In stock
+	 * - Out of stock
+	 * - Available on backorder
+	 * - 2 left in stock
+	 *
+	 * @param [bool]     $is_in_stock Whether the product is in stock.
+	 * @param [bool]     $is_low_stock Whether the product is low in stock.
+	 * @param [int|null] $low_stock_amount The amount of stock that is considered low.
+	 * @param [mixed]    $is_on_backorder Whether the product is on backorder.
+	 * @return string Stock text.
+	 */
+	protected static function getTextBasedOnStock( $is_in_stock, $is_low_stock, $low_stock_amount, $is_on_backorder ) {
+		if ( $is_low_stock ) {
+			return sprintf(
+				/* translators: %d is number of items in stock for product */
+				__( '%d left in stock', 'woo-gutenberg-products-block' ),
+				$low_stock_amount
+			);
+		} elseif ( $is_on_backorder ) {
+			return __( 'Available on backorder', 'woo-gutenberg-products-block' );
+		} elseif ( $is_in_stock ) {
+			return __( 'In stock', 'woo-gutenberg-products-block' );
+		} else {
+			return __( 'Out of stock', 'woo-gutenberg-products-block' );
+		}
+	}
+
+
+
+	/**
+	 * Include and render the block.
+	 *
+	 * @param array    $attributes Block attributes. Default empty array.
+	 * @param string   $content    Block content. Default empty string.
+	 * @param WP_Block $block      Block instance.
+	 * @return string Rendered block type output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		if ( ! empty( $content ) ) {
+			parent::register_block_type_assets();
+			$this->register_chunk_translations( [ $this->block_name ] );
+			return $content;
+		}
+
+		$post_id         = $block->context['postId'];
+		$product         = wc_get_product( $post_id );
+		$is_in_stock     = $product->is_in_stock();
+		$is_on_backorder = $product->is_on_backorder();
+
+		$low_stock_amount = $product->get_low_stock_amount();
+		$total_stock      = $product->get_stock_quantity();
+		$is_low_stock     = $low_stock_amount && $total_stock <= $low_stock_amount;
+
+		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
+
+		$classnames  = isset( $classes_and_styles['classes'] ) ? ' ' . esc_attr( $classes_and_styles['classes'] ) . ' ' : '';
+		$classnames .= isset( $attributes['className'] ) ? ' ' . esc_attr( $attributes['className'] ) . ' ' : '';
+		$classnames .= ! $is_in_stock ? ' wc-block-components-product-stock-indicator--out-of-stock ' : '';
+		$classnames .= $is_in_stock ? ' wc-block-components-product-stock-indicator--in-stock ' : '';
+		$classnames .= $is_low_stock ? ' wc-block-components-product-stock-indicator--low-stock ' : '';
+		$classnames .= $is_on_backorder ? ' wc-block-components-product-stock-indicator--available-on-backorder ' : '';
+
+		$output  = '';
+		$output .= '<div class="wc-block-components-product-stock-indicator ' . $classnames . '"';
```

Comment:
Personally, I would prefer to use the `esc_attr()` here than in L120 and L121. Escaping the data right before printing it helps making sure it's escaped; otherwise somebody reviewing the code needs to look at several lines up to see if it's escaped or not.

Reply:
I agree with you. Thanks for the suggestion. I will make the changes accordingly 🙂